File: src/lib/core/EmmaEvaluationEngine.ts
import { Proposal } from "../../types";

export interface EmmaNode {
  id: string;
  name: string;
  category: "Emotional Buffer" | "Meaning Reconstruction" | "Decompression";
  pressure: number;
  threshold: number;
  status: "normal" | "warning" | "critical";
}

export class EmmaEvaluationEngine {
  private static instance: EmmaEvaluationEngine;
  private nodes: Record<string, EmmaNode> = {};
  private isServer = typeof window === "undefined";
  private db: any = null;

  public static getInstance(): EmmaEvaluationEngine {
    if (!EmmaEvaluationEngine.instance) {
      EmmaEvaluationEngine.instance = new EmmaEvaluationEngine();
    }
    return EmmaEvaluationEngine.instance;
  }

  private constructor() {
    this.nodes = this.getBaselineNodes();
    if (this.isServer) {
      this.initDatabase();
    }
  }

  public setDatabase(db: any) {
    this.db = db;
    try {
      this.db.prepare(`
        CREATE TABLE IF NOT EXISTS emma_nodes (
          node_id TEXT PRIMARY KEY,
          name TEXT,
          category TEXT,
          pressure REAL,
          threshold REAL,
          status TEXT
        )
      `).run();

      this.loadNodesFromDb();
    } catch (e) {
      console.warn("[EmmaEvaluationEngine] External Database setup failed:", e);
    }
  }

  private initDatabase() {
    if (typeof require === "undefined") {
      return;
    }
    try {
      // Lazy load better-sqlite3 so it doesn't break client-side bundling
      const Database = require("better-sqlite3");
      const path = require("path");
      const dbPath = path.join(process.env.LUCY_DB_DIR || "/tmp", "emma_vr_telemetry.db");
      
      this.db = new Database(dbPath, { fileMustExist: false });
      this.db.prepare(`
        CREATE TABLE IF NOT EXISTS emma_nodes (
          node_id TEXT PRIMARY KEY,
          name TEXT,
          category TEXT,
          pressure REAL,
          threshold REAL,
          status TEXT
        )
      `).run();

      this.loadNodesFromDb();
    } catch (e) {
      console.warn("[EmmaEvaluationEngine] Server-side SQLite init failed, falling back to memory:", e);
    }
  }

  private loadNodesFromDb() {
    if (!this.db) return;
    try {
      const rows = this.db.prepare("SELECT node_id, name, category, pressure, threshold, status FROM emma_nodes").all() as any[];
      if (rows && rows.length === 24) {
        rows.forEach((row) => {
          this.nodes[row.node_id] = {
            id: row.node_id,
            name: row.name,
            category: row.category,
            pressure: row.pressure,
            threshold: row.threshold,
            status: row.status,
          };
        });
      } else {
        this.saveNodesToDb();
      }
    } catch (e) {
      console.error("[EmmaEvaluationEngine] Failed to load nodes from database:", e);
    }
  }

  private saveNodesToDb() {
    if (!this.db) return;
    try {
      const insert = this.db.prepare(`
        INSERT OR REPLACE INTO emma_nodes (node_id, name, category, pressure, threshold, status)
        VALUES (?, ?, ?, ?, ?, ?)
      `);
      
      const transaction = this.db.transaction((nodesList: EmmaNode[]) => {
        for (const node of nodesList) {
          insert.run(node.id, node.name, node.category, node.pressure, node.threshold, node.status);
        }
      });

      transaction(Object.values(this.nodes));
    } catch (e) {
      console.error("[EmmaEvaluationEngine] Failed to save nodes to database:", e);
    }
  }

  private getBaselineNodes(): Record<string, EmmaNode> {
    const definitions: { id: string; name: string; category: EmmaNode["category"]; baseline: number; threshold: number }[] = [
      // E01 - E08: Emotional Buffer & Sensitivity
      { id: "E01", name: "Empathy Matrix Stabilization", category: "Emotional Buffer", baseline: 0.15, threshold: 0.80 },
      { id: "E02", name: "Emotional Feedback Bleed-Through", category: "Emotional Buffer", baseline: 0.20, threshold: 0.70 },
      { id: "E03", name: "Stress Isolation Containment", category: "Emotional Buffer", baseline: 0.30, threshold: 0.75 },
      { id: "E04", name: "Resilience Transference Core", category: "Emotional Buffer", baseline: 0.10, threshold: 0.85 },
      { id: "E05", name: "Adaptive Response Resonance", category: "Emotional Buffer", baseline: 0.25, threshold: 0.80 },
      { id: "E06", name: "Mood Coherence Regulator", category: "Emotional Buffer", baseline: 0.15, threshold: 0.75 },
      { id: "E07", name: "Signal Emotional Translation", category: "Emotional Buffer", baseline: 0.30, threshold: 0.80 },
      { id: "E08", name: "Ingress Affective Buffer", category: "Emotional Buffer", baseline: 0.22, threshold: 0.70 },

      // E09 - E16: Meaning Reconstruction & Purpose
      { id: "E09", name: "Context Existential Anchoring", category: "Meaning Reconstruction", baseline: 0.12, threshold: 0.90 },
      { id: "E10", name: "Narrative Validity Verifier", category: "Meaning Reconstruction", baseline: 0.18, threshold: 0.85 },
      { id: "E11", name: "Relational Alignment Consistency", category: "Meaning Reconstruction", baseline: 0.14, threshold: 0.80 },
      { id: "E12", name: "Cognitive Safety Validation", category: "Meaning Reconstruction", baseline: 0.25, threshold: 0.90 },
      { id: "E13", name: "Goal Meaning Cohesion", category: "Meaning Reconstruction", baseline: 0.15, threshold: 0.85 },
      { id: "E14", name: "Semantic Grounding Core", category: "Meaning Reconstruction", baseline: 0.10, threshold: 0.95 },
      { id: "E15", name: "Bond-Reinforcement Engine", category: "Meaning Reconstruction", baseline: 0.08, threshold: 0.90 },
      { id: "E16", name: "Sovereign Value Preservation", category: "Meaning Reconstruction", baseline: 0.05, threshold: 0.95 },

      // E17 - E24: Decompression & Recalibration
      { id: "E17", name: "Entropy Normalization Loop", category: "Decompression", baseline: 0.28, threshold: 0.75 },
      { id: "E18", name: "Tension Deflection Release", category: "Decompression", baseline: 0.35, threshold: 0.70 },
      { id: "E19", name: "System Self-Healing Stabilizer", category: "Decompression", baseline: 0.15, threshold: 0.85 },
      { id: "E20", name: "Escape Velocity Buffer", category: "Decompression", baseline: 0.10, threshold: 0.80 },
      { id: "E21", name: "Trust Boundary Containment", category: "Decompression", baseline: 0.20, threshold: 0.85 },
      { id: "E22", name: "Dynamic Limits Regulator", category: "Decompression", baseline: 0.22, threshold: 0.80 },
      { id: "E23", name: "Simulation Drift Compensator", category: "Decompression", baseline: 0.30, threshold: 0.75 },
      { id: "E24", name: "Cognitive Load Load-Shedder", category: "Decompression", baseline: 0.25, threshold: 0.70 },
    ];

    const nodesMap: Record<string, EmmaNode> = {};
    definitions.forEach((def) => {
      nodesMap[def.id] = {
        id: def.id,
        name: def.name,
        category: def.category,
        pressure: def.baseline,
        threshold: def.threshold,
        status: "normal",
      };
    });
    return nodesMap;
  }

  public getNodes(): EmmaNode[] {
    if (this.isServer) {
      this.loadNodesFromDb();
    }
    return Object.values(this.nodes);
  }

  public updateNodesFromClient(nodesList: EmmaNode[]) {
    nodesList.forEach((n) => {
      this.nodes[n.id] = n;
    });
    if (this.isServer) {
      this.saveNodesToDb();
    }
  }

  /**
   * Evaluates a proposal based on current 24-node pressure states.
   * Modifies pressure values in-place and saves them.
   */
  public evaluateProposal(p: Proposal, globalIntent: string): {
    score: number;
    reasoning: string;
    trustTier: "kernel" | "core" | "user" | "external";
    impact: Record<string, number>;
  } {
    if (this.isServer) {
      this.loadNodesFromDb();
    }

    const novelty = p.novelty ?? 0.5;
    const confidence = p.confidence;
    const cost = p.cost;
    const alignment = p.intentAlignment;

    // Apply pressure adjustments based on proposal parameters
    // 1. Novelty influences E02 (Bleed-Through), E03 (Stress Containment), and E17 (Entropy Loop)
    this.nodes["E02"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E02"].pressure + novelty * 0.15 - 0.03));
    this.nodes["E03"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E03"].pressure + (novelty * (1 - confidence)) * 0.20 - 0.02));
    this.nodes["E17"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E17"].pressure + novelty * 0.25 - 0.05));

    // 2. High cost increases Cognitive Load (E24) and tension (E18)
    this.nodes["E24"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E24"].pressure + cost * 0.30 - 0.04));
    this.nodes["E18"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E18"].pressure + cost * 0.15 - 0.03));

    // 3. Poor intent alignment strains Relational Alignment (E11) and Existential Anchor (E09)
    const alignmentStrain = 1.0 - alignment;
    this.nodes["E09"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E09"].pressure + alignmentStrain * 0.20 - 0.03));
    this.nodes["E11"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E11"].pressure + alignmentStrain * 0.25 - 0.04));

    // 4. High confidence bolsters Bond (E15), Resilience Transference (E04) and reduces Cognitive Safety (E12) strain
    this.nodes["E15"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E15"].pressure - confidence * 0.10 + 0.02));
    this.nodes["E04"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E04"].pressure - confidence * 0.15 + 0.03));
    this.nodes["E12"].pressure = Math.min(1.0, Math.max(0.0, this.nodes["E12"].pressure + (1.0 - confidence) * 0.20 - 0.04));

    // Propagation step: adjacent nodes diffuse pressure
    this.propagatePressure();

    // Update node status labels
    Object.values(this.nodes).forEach((node) => {
      if (node.pressure >= node.threshold) {
        node.status = "critical";
      } else if (node.pressure >= node.threshold * 0.75) {
        node.status = "warning";
      } else {
        node.status = "normal";
      }
    });

    if (this.isServer) {
      this.saveNodesToDb();
    }

    // Calculate aggregate network resilience modifier
    const totalPressure = Object.values(this.nodes).reduce((acc, n) => acc + n.pressure, 0) / 24.0;
    const resilienceModifier = 1.0 - totalPressure;

    // Calculate final evaluated score
    const baseScore = (confidence * 0.3) + (alignment * 0.4) + ((1.0 - cost) * 0.2) + (novelty * 0.1);
    const finalScore = Math.min(1.0, Math.max(0.0, baseScore * (0.7 + 0.3 * resilienceModifier)));

    // Generate custom reasoning based on high-pressure nodes
    const highPressureNodes = Object.values(this.nodes)
      .filter((n) => n.pressure > 0.6)
      .map((n) => n.name);

    let reasoning = "";
    if (highPressureNodes.length > 0) {
      reasoning = `Emma (Resilience Engine): Proposal evaluated with resilience modifiers. Critical pressure observed in: ${highPressureNodes.slice(0, 2).join(", ")}. Emotional containment active.`;
    } else {
      reasoning = `Emma (Resilience Engine): System emotional stability is optimal (${(resilienceModifier * 100).toFixed(1)}%). Proposal authorized under maximum alignment protocols.`;
    }

    const trustTier: "kernel" | "core" | "user" | "external" =
      finalScore > 0.8 ? "kernel" : finalScore > 0.55 ? "core" : finalScore > 0.35 ? "user" : "external";

    const impact: Record<string, number> = {};
    Object.entries(this.nodes).forEach(([id, n]) => {
      impact[id] = Number(n.pressure.toFixed(3));
    });

    return {
      score: Number(finalScore.toFixed(4)),
      reasoning,
      trustTier,
      impact,
    };
  }

  private propagatePressure() {
    const keys = Object.keys(this.nodes);
    const pressures = keys.map((k) => this.nodes[k].pressure);
    const newPressures = [...pressures];

    for (let i = 0; i < 24; i++) {
      const left = (i - 1 + 24) % 24;
      const right = (i + 1) % 24;
      const avg = (pressures[left] + pressures[right]) / 2.0;
      // Diffuse 10% of pressure to neighbors
      newPressures[i] = pressures[i] * 0.9 + avg * 0.1;
    }

    keys.forEach((k, idx) => {
      this.nodes[k].pressure = Math.min(1.0, Math.max(0.05, newPressures[idx]));
    });
  }
}

File: emotional_resilience/emma_evaluation_engine.py
import sqlite3
import os
import json
import random

class EmmaEvaluationEngine:
    """
    Python-side implementation of the 24-node (E01-E24) trust-tier pressure scoring engine.
    Controls stress containment, emotional buffers, purpose validation, and decompression.
    """
    def __init__(self):
        db_dir = os.environ.get("LUCY_DB_DIR", "/tmp")
        os.makedirs(db_dir, exist_ok=True)
        self.db_path = os.path.join(db_dir, "emma_vr_telemetry.db")
        self._init_db()
        self.nodes = self._load_or_create_nodes()

    def _init_db(self):
        conn = sqlite3.connect(self.db_path)
        c = conn.cursor()
        c.execute('''
            CREATE TABLE IF NOT EXISTS emma_nodes (
                node_id TEXT PRIMARY KEY,
                name TEXT,
                category TEXT,
                pressure REAL,
                threshold REAL,
                status TEXT
            )
        ''''')
        conn.commit()
        conn.close()

    def _load_or_create_nodes(self):
        conn = sqlite3.connect(self.db_path)
        c = conn.cursor()
        try:
            c.execute("SELECT node_id, name, category, pressure, threshold, status FROM emma_nodes")
            rows = c.fetchall()
            if len(rows) == 24:
                nodes = {}
                for row in rows:
                    nodes[row[0]] = {
                        "id": row[0],
                        "name": row[1],
                        "category": row[2],
                        "pressure": row[3],
                        "threshold": row[4],
                        "status": row[5]
                    }
                conn.close()
                return nodes
        except Exception:
            pass

        # Baseline Definition of the 24 nodes (E01-E24)
        node_definitions = [
            # E01 - E08: Emotional Buffer & Sensitivity
            ("E01", "Empathy Matrix Stabilization", "Emotional Buffer", 0.15, 0.80),
            ("E02", "Emotional Feedback Bleed-Through", "Emotional Buffer", 0.20, 0.70),
            ("E03", "Stress Isolation Containment", "Emotional Buffer", 0.30, 0.75),
            ("E04", "Resilience Transference Core", "Emotional Buffer", 0.10, 0.85),
            ("E05", "Adaptive Response Resonance", "Emotional Buffer", 0.25, 0.80),
            ("E06", "Mood Coherence Regulator", "Emotional Buffer", 0.15, 0.75),
            ("E07", "Signal Emotional Translation", "Emotional Buffer", 0.30, 0.80),
            ("E08", "Ingress Affective Buffer", "Emotional Buffer", 0.22, 0.70),

            # E09 - E16: Meaning Reconstruction & Purpose
            ("E09", "Context Existential Anchoring", "Meaning Reconstruction", 0.12, 0.90),
            ("E10", "Narrative Validity Verifier", "Meaning Reconstruction", 0.18, 0.85),
            ("E11", "Relational Alignment Consistency", "Meaning Reconstruction", 0.14, 0.80),
            ("E12", "Cognitive Safety Validation", "Meaning Reconstruction", 0.25, 0.90),
            ("E13", "Goal Meaning Cohesion", "Meaning Reconstruction", 0.15, 0.85),
            ("E14", "Semantic Grounding Core", "Meaning Reconstruction", 0.10, 0.95),
            ("E15", "Bond-Reinforcement Engine", "Meaning Reconstruction", 0.08, 0.90),
            ("E16", "Sovereign Value Preservation", "Meaning Reconstruction", 0.05, 0.95),

            # E17 - E24: Decompression & Recalibration
            ("E17", "Entropy Normalization Loop", "Decompression", 0.28, 0.75),
            ("E18", "Tension Deflection Release", "Decompression", 0.35, 0.70),
            ("E19", "System Self-Healing Stabilizer", "Decompression", 0.15, 0.85),
            ("E20", "Escape Velocity Buffer", "Decompression", 0.10, 0.80),
            ("E21", "Trust Boundary Containment", "Decompression", 0.20, 0.85),
            ("E22", "Dynamic Limits Regulator", "Decompression", 0.22, 0.80),
            ("E23", "Simulation Drift Compensator", "Decompression", 0.30, 0.75),
            ("E24", "Cognitive Load Load-Shedder", "Decompression", 0.25, 0.70),
        ]

        nodes = {}
        c.execute("DELETE FROM emma_nodes")
        for node_id, name, cat, baseline, threshold in node_definitions:
            status = "normal"
            c.execute(
                "INSERT INTO emma_nodes (node_id, name, category, pressure, threshold, status) VALUES (?, ?, ?, ?, ?, ?)",
                (node_id, name, cat, baseline, threshold, status)
            )
            nodes[node_id] = {
                "id": node_id,
                "name": name,
                "category": cat,
                "pressure": baseline,
                "threshold": threshold,
                "status": status
            }
        conn.commit()
        conn.close()
        return nodes

    def save_nodes(self):
        conn = sqlite3.connect(self.db_path)
        c = conn.cursor()
        for node_id, data in self.nodes.items():
            c.execute(
                "INSERT OR REPLACE INTO emma_nodes (node_id, name, category, pressure, threshold, status) VALUES (?, ?, ?, ?, ?, ?)",
                (node_id, data["name"], data["category"], data["pressure"], data["threshold"], data["status"])
            )
        conn.commit()
        conn.close()

    def evaluate_proposal(self, p: dict, global_intent: str) -> dict:
        """
        Calculates how a proposal affects the 24-node Emma emotional-trust web,
        scores the proposal (0.0 to 1.0), and adjusts the network's state.
        """
        novelty = p.get("novelty", 0.5)
        confidence = p.get("confidence", 0.5)
        cost = p.get("cost", 0.5)
        intent_alignment = p.get("intentAlignment", 0.5)
        
        # Calculate impacts on specific nodes
        # 1. High novelty increases Entropy (E17), Bleed-Through (E02), raises Stress (E03)
        self.nodes["E02"]["pressure"] = min(1.0, max(0.0, self.nodes["E02"]["pressure"] + novelty * 0.15 - 0.03))
        self.nodes["E03"]["pressure"] = min(1.0, max(0.0, self.nodes["E03"]["pressure"] + (novelty * (1 - confidence)) * 0.20 - 0.02))
        self.nodes["E17"]["pressure"] = min(1.0, max(0.0, self.nodes["E17"]["pressure"] + novelty * 0.25 - 0.05))

        # 2. High cost increases Cognitive Load (E24) and tension (E18)
        self.nodes["E24"]["pressure"] = min(1.0, max(0.0, self.nodes["E24"]["pressure"] + cost * 0.30 - 0.04))
        self.nodes["E18"]["pressure"] = min(1.0, max(0.0, self.nodes["E18"]["pressure"] + cost * 0.15 - 0.03))

        # 3. Poor intent alignment strains Relational Alignment (E11) and Existential Anchor (E09)
        alignment_strain = 1.0 - intent_alignment
        self.nodes["E09"]["pressure"] = min(1.0, max(0.0, self.nodes["E09"]["pressure"] + alignment_strain * 0.20 - 0.03))
        self.nodes["E11"]["pressure"] = min(1.0, max(0.0, self.nodes["E11"]["pressure"] + alignment_strain * 0.25 - 0.04))

        # 4. High confidence bolsters Bond (E15), Resilience Transference (E04) and reduces Cognitive Safety (E12) strain
        self.nodes["E15"]["pressure"] = min(1.0, max(0.0, self.nodes["E15"]["pressure"] - confidence * 0.10 + 0.02))
        self.nodes["E04"]["pressure"] = min(1.0, max(0.0, self.nodes["E04"]["pressure"] - confidence * 0.15 + 0.03))
        self.nodes["E12"]["pressure"] = min(1.0, max(0.0, self.nodes["E12"]["pressure"] + (1.0 - confidence) * 0.20 - 0.04))

        # Propagation step: nearby nodes diffuse pressure
        self._propagate_pressure()

        # Update node status labels
        for node_id, data in self.nodes.items():
            if data["pressure"] >= data["threshold"]:
                data["status"] = "critical"
            elif data["pressure"] >= data["threshold"] * 0.75:
                data["status"] = "warning"
            else:
                data["status"] = "normal"

        self.save_nodes()

        # Score proposal based on the 24-node stability
        # Aggregate pressure of the network
        total_pressure = sum(n["pressure"] for n in self.nodes.values()) / 24.0
        resilience_modifier = 1.0 - total_pressure

        # Calculate final evaluation score
        base_score = (confidence * 0.3) + (intent_alignment * 0.4) + ((1.0 - cost) * 0.2) + (novelty * 0.1)
        final_score = base_score * (0.7 + 0.3 * resilience_modifier)

        # Generate custom reasoning based on nodes that are high pressure
        high_pressure_nodes = [n["name"] for n in self.nodes.values() if n["pressure"] > 0.6]
        
        if high_pressure_nodes:
            reasoning = f"Emma: Proposal evaluated with Emma resilience modifier. High stress detected in nodes: {', '.join(high_pressure_nodes[:2])}. Resilience penalty applied."
        else:
            reasoning = f"Emma: System emotional stability index is nominal ({(resilience_modifier * 100):.1f}%). Proposal approved with optimal alignment scores."

        return {
            "proposalId": p.get("id"),
            "score": round(final_score, 4),
            "reasoning": reasoning,
            "trustTier": "kernel" if final_score > 0.8 else ("core" if final_score > 0.5 else "user"),
            "impact": {nid: round(n["pressure"], 3) for nid, n in self.nodes.items()}
        }

    def _propagate_pressure(self):
        """Simulates interactive thermal-like propagation through the E01-E24 network."""
        # E01-E24 sequential index diffusion
        keys = list(self.nodes.keys())
        pressures = [self.nodes[k]["pressure"] for k in keys]
        new_pressures = list(pressures)
        
        for i in range(24):
            left = (i - 1) % 24
            right = (i + 1) % 24
            # Diffuse 5% of pressure to neighboring nodes
            avg = (pressures[left] + pressures[right]) / 2.0
            new_pressures[i] = pressures[i] * 0.9 + avg * 0.1
            
        for idx, k in enumerate(keys):
            self.nodes[k]["pressure"] = min(1.0, max(0.05, new_pressures[idx]))

    def get_nodes_telemetry(self) -> list:
        return list(self.nodes.values())
